### PR TITLE
Fix mods wrongly adding audio adjustments after being toggled off

### LINKED ISSUE #18178: Deselecting Wind Up/Down mod when pitch adjustment is turned off results in it turning back on
### Type

Game behaviour

### Bug description

simple to reproduce simply select the wind up/down mod then turn off the pitch adjustment and deselect the mod

### Screenshots or videos

https://user-images.githubusercontent.com/43358824/167393318-b26d97b6-f3c3-4eb6-9b7d-6c70a13ca051.mp4

### Version

2022.509.0

### Logs

[database.log](https://github.com/ppy/osu/files/8650888/database.log)
[network.log](https://github.com/ppy/osu/files/8650889/network.log)
[performance.log](https://github.com/ppy/osu/files/8650890/performance.log)
[runtime.log](https://github.com/ppy/osu/files/8650891/runtime.log)
[updater.log](https://github.com/ppy/osu/files/8650892/updater.log)

### PR DESCRIPTION
**Pull request by bdach**
Closes #18178.

First, an explanation why this broke. On master, `ModTimeRamp` stores a reference to the last track supplied on `IApplicableToTrack`:

https://github.com/ppy/osu/blob/5b17e92770e1613d099599b9fc8f758a10b0c479/osu.Game/Rulesets/Mods/ModTimeRamp.cs#L60

It then uses this track reference to access `CurrentTime` (which is not the problem), but also to dynamically add/remove audio adjustments when the "Adjust Pitch" option is toggled (which is the problem):

https://github.com/ppy/osu/blob/5b17e92770e1613d099599b9fc8f758a10b0c479/osu.Game/Rulesets/Mods/ModTimeRamp.cs#L101-L107

It turns out that this only worked correctly on the old overlay due to a particular ordering of operations. With the old mod overlay, the following happened:

* On deselection, `ModButton` [resets settings to defaults first](https://github.com/ppy/osu/blob/5b17e92770e1613d099599b9fc8f758a10b0c479/osu.Game/Overlays/Mods/ModButton.cs#L74-L75)
  * The deselection potentially triggers a change of `AdjustPitch` to defaults, therefore potentially swapping a tempo adjustment for a frequency adjustment
* `ModButton` then [signals that its selection has changed](https://github.com/ppy/osu/blob/5b17e92770e1613d099599b9fc8f758a10b0c479/osu.Game/Overlays/Mods/ModButton.cs#L111), eventually triggering a game-global `SelectedMods` update
* `MusicController` [subscribes to mod changes](https://github.com/ppy/osu/blob/5b17e92770e1613d099599b9fc8f758a10b0c479/osu.Game/Overlays/MusicController.cs#L64) and [resets adjustments](https://github.com/ppy/osu/blob/5b17e92770e1613d099599b9fc8f758a10b0c479/osu.Game/Overlays/MusicController.cs#L64) after all that.

On the new design, the ordering there ends up basically swapped - the adjustments are reset, the mod is reset to defaults after that, which causes an `AdjustPitch` change, which causes a dynamic adjustment to be added and triggers the bug.

Rather than focus on which ordering is correct, I initially resolved to add an `IApplicableToTrack.UnapplyFromTrack()` method, but then figured that this isn't really required so long as mods stop adding adjustments on held references to tracks. Which is what this pull does - `ModTimeRamp` and `ModAdaptiveSpeed` are changed so that they don't apply adjustments to held references to tracks, and ample warning is added to xmldoc to not do this. Unfortunately in `ModTimeRamp` I cannot fully get rid of the reference, as that mod also accesses `ITrack.CurrentTime`.

I know that @frenzibyte was really adamant about [some fool-proof way to add adjustments](https://discord.com/channels/188630481301012481/188630652340404224/973314948144971847) that would make it impossible for this to ever happen, but (a) my brain somehow just stops working when reading any of that proposal, and (b) it doesn't solve the fact that `ModTimeRamp` needs `ITrack` to work, and at the point where you've got a track reference in hand again, you can bug out again.

### CODE REVIEW

**Comment by frenzibyte** [comment]
I'm just going to note that I'm personally fine with either solutions, but I thought it may be better going the fool-proof way as it wouldn't require that many lines of changes (maybe, maybe not...). 

I've looked into it and this is pretty much what I imagine to be done (https://github.com/ppy/osu/commit/708fc471f4af6894e2b94c50a2d5c524069a9e63), with the second point in OP being fixed by using `playfield.Clock.CurrentTime` instead of `track.CurrentTime`.

Unfortunately, there is still some breakage going around with it and I don't have the time to look into that yet (for one, go to player with wind up, let the track start accelerating, exit to song select and deselect wind up. song should reset adjustment but it doesn't). So only bringing this comment up for clarification and to show the LoCs necessary to *mostly* achieve this.

**Comment by smoogipoo** [comment]
I'm not against this, but I would also like to explore @frenzibyte 's version in the future, since I think that reads better.

**Review by smoogipoo — DISMISSED** [review]

**Review by peppy — DISMISSED** [review]

**Comment by peppy** [comment]
Honestly I was going to just let this one fly without worrying too much about the implementation, but I think @frenzibyte's proposal holds out better. At least I can understand the code by reading it, while this PR's unbind logic is a bit... hard to follow.

**Comment by bdach** [comment]
I have... several doubts about the correctness of the alternative, on a five second read. But whatever, if it gets PRed then I'll say my piece about it and leave it be, and you can decide which one you prefer.

**Comment by bdach** [comment]
Closing as no longer required.
